str, htc, htp: Change default materialization placeholder type from `Unit` to `Any`

sirthias opened the pull request:
Closes #17039 and #16933.

I needed this a basis for the current work in HTTP so I simply got it done altogether.
(For making things a bit easier for me) this PR builds on #17126, so please review the first 3 commits there. Also #17126 should be merged first.

sirthias committed:
+str Add `FanInShape1N`

sirthias committed:
!htc Some cleanup in internal and external naming, improvements to scaladoc

sirthias committed:
=htc upgrade HTTP layer API to new BidiFlow design

sirthias committed:
!str #17039 change default materialization placeholder type from `Unit` to `Any`

sirthias committed:
!htc #17039, #16933 change default materialization placeholder type from `Unit` to `Any`

sirthias committed:
=doc #17039 change default materialization placeholder type from `Unit` to `Any`

drewhk commented on akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeDigest.scala:
```diff
@@ -41,7 +41,7 @@ class RecipeDigest extends RecipeSpec {
         }
       }
 
-      val digest: Source[ByteString, Unit] = data.transform(() => digestCalculator("SHA-256"))
```
Nope. This is proper Unit.

drewhk commented:
This is changing too much. Things should *return* Unit as the materialized type if they really do not materialize anything meaningful. Only those parts should be changed where we *accept* something with an unknown materialized type. So it should be

 - Unit if *we* return it and we know it does not have a materialized value
 - Any if we *take it* as a parameter and we do not care about the materialized value

ktoso commented:
Meh, @drewhk was faster this time :-) 
It's as Endre writes, in `IncommingConnection(..., flow: Flow[ByteString, ByteString, Any]) {` we really want to accept Any, but when returning things we should expose the most specific type - Unit signalling nothing is going to be materialized.

sirthias commented:
Ok, agreed. I'll get this fixed up ASAP.

jrudolph commented:
+1 on doing this soon. It would be good if I could build on this rather sooner than later.

sirthias commented:
I'm on it and almost done.
Let me say this: getting this new type parameter right every (especially with regard to Java interop) is tricky (but doable).

sirthias committed:
+str #17039 add `concatMat` to `Flow` and `Source`

sirthias commented:
Ok, applied the discussed reduction in scope.
In the end, not that much needs to be changed.

Apart from some smaller cleanups, the changes in the akka-stream module are essentially:

- Make `Flow::section` and `Source::section` accept function widened from Mat `Unit` to Mat `Any`.
  Also change it from `Keep.right` semantics to `Keep.left`, as the latter is the much more frequently required case.

- Make `FlattenStrategy` covariant in its second type param. Unfortunately, on the Java side, `FlattenStrategy.concat` needed to receive a second type param.

- Added `concatMat` to `Flow` and `Source` in order to solve the original problem reported by @RichardBradley in #17039

- While I was at it I also changed
  ```scala
  Source.apply[T](publisher: Publisher[T]): Source[T, Unit]
  ```
  to
  ```scala
  Source.apply[T](publisher: Publisher[_ >: T]): Source[T, Unit]
  ```
  (as well as, accordingly, the respective constructor on `Sink`). As this appears to be an (initially?) obvious thing to do there might be a reason why this hasn't been done so far. Please let me know if this is the case.

drewhk commented:
Cool! Actually section will probably go away now, since you can do the following now:

    val sectionFlow = Flow[X].map(f).withAttributes(attr)
    val parentFlow = Flow[Y].mapAsync(ff).via(sectionFlow).map(f2).filter(p) ...

all thanks to the now hierarchic internal layout representation.

rkuhn commented on akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeWorkerPool.scala:
```diff
@@ -17,7 +17,7 @@ class RecipeWorkerPool extends RecipeSpec {
 
       //#worker-pool
-      def balancer[In, Out](worker: Flow[In, Out, Unit], workerCount: Int): Flow[In, Out, Unit] = {
+      def balancer[In, Out, Mat](worker: Flow[In, Out, Any], workerCount: Int): Flow[In, Out, Unit] = {
```
this type parameter is not used

sirthias replied:
Indeed. Left-over from some previous iteration.
Will remove.

rkuhn commented on akka-http-core/src/main/scala/akka/http/Http.scala:
```diff
@@ -194,7 +200,7 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
     localAddress: InetSocketAddress,
     remoteAddress: InetSocketAddress,
-    flow: Flow[HttpResponse, HttpRequest, Unit]) {
+    flow: Flow[HttpResponse, HttpRequest, Any]) {
```
here we provide a flow to users, so `Unit` was the right type

sirthias replied:
Yup.

rkuhn commented on akka-http-core/src/main/scala/akka/http/Http.scala:
```diff
@@ -128,44 +135,44 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+  def clientLayer(remoteAddress: InetSocketAddress, // TODO: remove after #16168 is cleared
+                  settings: ClientConnectionSettings,
+                  log: LoggingAdapter = system.log): ClientLayer =
+    BidiFlow.wrap(OutgoingConnectionBlueprint(remoteAddress, settings, log))
```
why does the factory not directly return a BidiFlow?

sirthias replied:
No reason. We could easily move the `BidiFlow.wrap` into `OutgoingConnectionBlueprint` if you'd prefer that.

rkuhn commented on akka-http-core/src/main/scala/akka/http/Http.scala:
```diff
@@ -128,44 +135,44 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
-                                           log: LoggingAdapter = system.log): Flow[HttpRequest, HttpResponse, Mat] = {
-    val effectiveSettings = ClientConnectionSettings(settings)
-    val clientBlueprint = HttpClient.clientBlueprint(remoteAddress, effectiveSettings, log)
+  type ClientLayer = BidiFlow[HttpRequest, ByteString, ByteString, HttpResponse, Any]
```
this is a return type that we provide to users, so it should be the concrete type that the blueprint materializes to

sirthias replied:
Yes, `Any` should be `Unit` here.
The client-side HTTP layer itself doesn't materialize to anything useful.

rkuhn commented on akka-http-core/src/main/scala/akka/http/Http.scala:
```diff
@@ -99,27 +99,34 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
-        import FlowGraph.Implicits._
-        server.httpRequests ~> user.inlet
-        user.outlet ~> server.httpResponses
+  type ServerLayer = BidiFlow[HttpResponse, ByteString, ByteString, HttpRequest, Any]
```
same as for ClientLayer

rkuhn commented on akka-http-core/src/main/scala/akka/http/Http.scala:
```diff
@@ -99,27 +99,34 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+   */
+  def serverLayer(settings: ServerSettings,
+                  log: LoggingAdapter = system.log)(implicit mat: FlowMaterializer): ServerLayer =
+    BidiFlow.wrap(HttpServerBluePrint(settings, log))
```
same as below

rkuhn commented on akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template:
```diff
@@ -57,6 +57,21 @@ class UniformFanInShape[T, O](val n: Int, _init: FanInShape.Init[O]) extends Fan
+    in1Vector(n - 1)
+  }
+}
+
```
you keep trying to sneak this in :-)

I think I know where it will be used, so leave it in this time.

sirthias replied:
I simply proposed it.
This is the first time you (or anyone else) made a statement as to whether to leave it in or not... :)

rkuhn commented on akka-stream/src/main/scala/akka/stream/javadsl/BidiFlow.scala:
```diff
@@ -12,7 +12,7 @@ object BidiFlow {
 
   /**
-   * A graph with the shape of a flow logically is a flow, this method makes
+   * A graph with the shape of a BidiFlow logically is a BidiFlow, this method makes
```
good catch!

rkuhn commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -58,7 +58,7 @@ object Source {
    * back-pressure upstream.
    */
-  def from[O](publisher: Publisher[O]): javadsl.Source[O, Unit] =
+  def from[O](publisher: Publisher[_ >: O]): javadsl.Source[O, Unit] =
```
This is wrong: a `Publisher[O]` is most adequately represented as a `Source[O]`. Then if a `Source[O2 >: O]` is required the Java way is to fix that at the use-site of the Source and not here. In any case, if we were to do something here then it would need to be `Publisher[_ <: O]` because otherwise our Source would sometimes emit things outside of its allowed type bound.

sirthias replied:
Of course. No idea what I was thinking here. I'll throw it out.

rkuhn commented on akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala:
```diff
@@ -57,7 +57,7 @@ object Sink extends SinkApply {
    * Helper to create [[Sink]] from `Subscriber`.
    */
-  def apply[T](subscriber: Subscriber[T]): Sink[T, Unit] =
+  def apply[T](subscriber: Subscriber[_ <: T]): Sink[T, Unit] =
```
same as above, with inverted polarity

rkuhn commented on akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:
```diff
@@ -171,7 +178,7 @@ object Source extends SourceApply {
    * back-pressure upstream.
    */
-  def apply[T](publisher: Publisher[T]): Source[T, Unit] =
+  def apply[T](publisher: Publisher[_ >: T]): Source[T, Unit] =
```
Here it is even more wrong, since Scala does have declaration-site variance, so this will never be useful (unless I am missing something).

sirthias replied:
Yeah, this is complete BS.
Sorry about that.

rkuhn replied:
no worries, sometimes we get carried away :-)

rkuhn commented on akka-http-core/src/main/scala/akka/http/model/HttpEntity.scala:
```diff
@@ -160,11 +160,11 @@ object HttpEntity {
     override def toStrict(timeout: FiniteDuration)(implicit fm: FlowMaterializer) =
```
line 130 is the most important one to fix

sirthias replied:
Yeah, but, AFAICS, the only way to do it is to add a `data.mapMaterialized(_ ⇒ ()))`, which incurs overhead *always*, even if I already pass in a `Source[ByteString, Unit]`.

rkuhn replied:
The solution is simple: use `Any` for the field’s type. Fixing this particular signature was the purpose of the whole exercise AFAIK, so I don’t think it is optional ;-)

sirthias replied:
Well, the member is at same time something we accept from the user as well as something that we supply. So we have conflicting goals of accepting `Any` but supplying `Unit`.
I'll make the change to `Any` here and in all other similar cases.

patriknw commented on akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template:
```diff
@@ -57,6 +57,21 @@ class UniformFanInShape[T, O](val n: Int, _init: FanInShape.Init[O]) extends Fan
+  override def deepCopy(): FanInShape1N[T0, T1, O] = super.deepCopy().asInstanceOf[FanInShape1N[T0, T1, O]]
+
+  val in0 = newInlet[T0]("in0")
+  val in1Vector: Vector[Inlet[T1]] = Vector.tabulate(n)(i => newInlet[T1](s"in${i+1}"))
```
We normally use `immutable.IndexedSeq` as type in public api. Is there a reason to use `Vector`?

sirthias replied:
I'll happily change it to `immutable.IndexedSeq`.

drewhk commented:
If I understand this also closes https://github.com/akka/akka/issues/16932

sirthias commented:
@drewhk Yes, it does. I'll make the change to `immutable.IndexedSeq` and rebase right now, so we can get this merged.

drewhk commented:
Yaay!

sirthias commented:
Done.

rkuhn commented:
well, I’ve got good news and bad news … the good is that I was about to click the button, the bad is that the button is not green ;-)

sirthias commented:
Damn. This thing really conflicts with everything.
I'll give it another try.

rkuhn commented:
quickly merged :-)

drewhk commented:
Yaay :)

jrudolph commented:
The only way to get this done!

sirthias commented:
Thank god!

rkuhn commented:
nah, was only me, nobody special …
